io/pci/pci_hotplug.py

```diff
@@ -124,20 +128,58 @@ def is_added():
-            False otherwise.
+            Checks if the block device adapter is recovers all its disks/paths
+            properly after hotplug of adapter.
+            Returns True if all disks/paths back online after adapter added
+            Back, else False.
             """
-            post = len(pci.get_disks_in_pci_address(pci_addr))
-            post += len(pci.get_nics_in_pci_address(pci_addr))
-            self.log.debug("Pre: %d,  Post: %d",
-                           self.end_devices[pci_addr], post)
-            if post == self.end_devices[pci_addr]:
+            def is_path_online():
+                path_stat = list(multipath.get_path_status(curr_path))
+                if path_stat[0] != 'active' or path_stat[2] != 'ready':
+                    return False
                 return True
-            return False
 
-        if not wait.wait_for(is_added, timeout=10):
+            curr_path = ''
+            err_disks = []
+            if pci.get_pci_class_name(pci_addr) == 'fc_host':
+                disks = pci.get_disks_in_pci_address(pci_addr)
+                for disk in disks:
+                    curr_path = disk.split("/")[-1]
+                    self.log.info("curr_path=%s" % curr_path)
+                    if not wait.wait_for(is_path_online, timeout=10):
+                        self.log.info("%s failed to recover after add" % disk)
+                        err_disks.append(disk)
+
+            if err_disks:
+                self.log.info("few paths failed to recover : %s" % err_disks)
+                return False
+            return True
+
+        def net_recovery_check():
+            """
+            Checks if the network adapter fuctionality like ping/link_state,
+            after adapter added back.
+            Returns True on propper Recovery, False if not.
+            """
+            self.log.info("entering the net recovery check")
+            local = LocalHost()
+            iface = pci.get_interfaces_in_pci_address(pci_addr, 'net')
+            networkinterface = NetworkInterface(iface[0], local)
+            if wait.wait_for(networkinterface.is_link_up, timeout=120):
+                if networkinterface.ping_check(self.peer_ip, count=5) is None:
+                    self.log.info("inteface is up and pinging")
+                    return True
+                return False
             return False
-        # Waiting for 10s per end device, for recovery.
-        return wait.wait_for(is_recovered,
-                             self.end_devices[pci_addr] * 10)
+
+        if wait.wait_for(is_added, timeout=30):
+            time.sleep(45)
+            if pci.get_pci_class_name(pci_addr) == 'net':
+                if wait.wait_for(net_recovery_check, timeout=30):
+                    return True
+                return False
+            else:
+                if wait.wait_for(is_recovered, timeout=30):
+                    return True
+                return False
```

Comment:
same here  keep single return false for overall hotplug_add function.... meaning if things works fine it returns true .. but any where if it fails,, it should come to end of hotplug_add and return False.. so no need to return false at every small function inside the hotplug_add